### PR TITLE
Add web sockets support

### LINKED ISSUE #117: Websockets built-in support
Hi, I got a use case where I have to put up a websockets server on top of PYLSP. 
Will there be any support for Websockets in future?

I'm trying to add one on my own atleast for my use-case but not able to understand the message flow structure. Can you provide help on how the messaging flow happens.

Main problem is: A/c to web sockets spec, there can be 'n' number of connections which can individually be a separate LSP session that means there should be 'n' PYLSP processes if using IO server. 

I'm trying to setup 'n' PYLSP objects in a single process for 'n' web sockets connections. Is there a document on how the 'class PythonLSPServer' works.

Thank you in advance.

### PR DESCRIPTION
**Pull request by npradeep357**
This pull request will enable PYLSP to run in Web Sockets configuration and can handle multiple connections in parallel. 
The Idea was to handle ~~multiple~~ parallel editing sessions using single process. This is achieved by creating new instances of PYLSP ~~class~~ Object for each WS connection in a single server process.

Fixes #117 

As of now this PR only supports text based messaging over WS. Binary messaging support will be added in future.

This PR uses ~~autobahn websockets with twisted framework~~ [Websockets Library](https://websockets.readthedocs.io/en/stable/)
~~There is asyncio based framework but twisted is used because PYLSP is based on callbacks and asyncio is not compatible with it.~~

referred docs: 
~~https://autobahn.readthedocs.io/en/latest/websocket/programming.html
https://autobahn.readthedocs.io/en/latest/installation.html~~
https://websockets.readthedocs.io/en/stable/

related dependencies are added to ~~setup.py~~ setup.cfg
related usage documentation is added to README

### CODE REVIEW

**Review by ccordoba12 — COMMENTED** [review]
Hey @npradeep357, thanks a lot for your contribution! My main comment for now is it's possible to add some tests for this new functionality, given that right now I see none.

Besides that, I'd like to ask @krassowski what he thinks about this one.

**Comment by npradeep357** [comment]
@ccordoba12, Thank you, will add tests for this.

Also, regarding the checks, there are Linux, Mac and Windows tests failure which says the version mentioned is not available for Py 3.6. I took the latest version of autobahn from PyPI which is [21.11.1](https://pypi.org/project/autobahn/).

But the log in the tests shows only until 21.2.1 which is an older version released in Feb 2021. ref: [autobahn PyPI history](https://pypi.org/project/autobahn/#history)

Suggest me if I have to move to older version or is there something I'm missing for Py 3.6?

**Comment by npradeep357** [comment]
After some digging, I found out that autobahn is only supported from Python 3.7+.

Is there a reason that PYLSP is supported from Python 3.6+? Can we move PYLSP to Python 3.7+? or should I downgrade the autobahn library which I used?

**Comment by ccordoba12** [comment]
> Is there a reason that PYLSP is supported from Python 3.6+?

There is because Python 3.6 is still a supported Python version. However, since it's going to reach end-of-life this month, we can drop support for it now.

Please do that in a different pull request though, i.e. increase the Python version in our `setup.py` and remove the Python 3.6 slots on Github actions. After that you could rebase this PR so that your tests pass.

**Comment by npradeep357** [comment]
Hey @ccordoba12, I have created a separate PR for version change as mentioned. Can you review and approve #130 when you find some time. Thank you

File: README.md
Comment (krassowski) [inline README.md]: Let's move this block to the end of `Configuration` section -  an average user does not care what method is used.

File: README.md
Comment (krassowski) [inline README.md]: I think we do not use PYLSP in capitals elsewhere so maybe use the full name?

Comment (krassowski) [reply README.md]:
```suggestion
Python LSP Server can communicate over WebSockets when configured as follows:
```

File: README.md
Comment (krassowski) [inline README.md]: Let's keep it simple and easier to maintain.

```suggestion
```

File: pylsp/python_lsp.py
Comment (krassowski) [inline pylsp/python_lsp.py]: You could wrap these with try-except and re-raise with instructions for installing websocket requirements, something like "Please run `pip install 'python-lsp-server[autobahn-twisted]` or equivalent".

File: setup.py
Comment (krassowski) [inline setup.py]: I wonder if the extras should be renamed to `websockets` so it is not tied to the current implementation if we need to change it in the future.

**Comment by npradeep357** [comment]
@krassowski made the suggested changes.

File: pylsp/python_lsp.py
Comment (krassowski) [inline pylsp/python_lsp.py]: Please remove commented code, or explain why it is commented.

File: pylsp/python_lsp.py
Comment (krassowski) [inline pylsp/python_lsp.py]:
```suggestion
        # if Consumer is None, it is assumed that the default streams-based approach is being used
```

**Comment by krassowski** [comment]
Thanks! Looks fine by me, only minor comments now.

**Review by krassowski — APPROVED** [review]
Broadly speaking it looks ok for me. Not sure about handling of the binary messages - should it emit a warning when one is received?

**Comment by npradeep357** [comment]
> Broadly speaking it looks ok for me. Not sure about handling of the binary messages - should it emit a warning when one is received?

Binary messages format, in general, will not be used in LSP use case as the protocol requires to send JSON messages over sockets. I assume there won't be a client which uses Binary format.

**Comment by npradeep357** [comment]
@krassowski, Not sure if this the right place, This PR requires #130 to be merged before merging this. Can you please review it and approve so that it can be merged before this.

Thank you.

**Comment by npradeep357** [comment]
@ccordoba12, can this also be added to v1.4.0 milestone?

**Comment by ccordoba12** [comment]
I'd still like to see some tests for this. See for instance:

https://stackoverflow.com/questions/40884504/pytest-testing-a-websocket-connection

**Comment by npradeep357** [comment]
Hi @ccordoba12, I'm still figuring out approach to complete the test case. I don't think this will go into v1.4.0. May can we move this out of the milestone for now?

**Comment by ccordoba12** [comment]
No problem. Hope you could resume this work soon.

File: setup.cfg
Comment (ccordoba12) [inline setup.cfg]: I suggest to remove this line because Twisted is a big dependency and it's not really necessary for the server to work. If people want ẁebsocket functionality, they can use the `[websocket]` extra requirement directly.

File: pylsp/python_lsp.py
Comment (ccordoba12) [inline pylsp/python_lsp.py]: Please move the `ujson` import to the beginning of the file because it's a main dependency of this project, so you don't need to check for its presence.

File: pylsp/python_lsp.py
Comment (ccordoba12) [inline pylsp/python_lsp.py]:
```suggestion
        # if consumer is None, it is assumed that the default streams-based approach is being used
```

File: README.md
Comment (ccordoba12) [inline README.md]:
```suggestion
The [autobahn-python](https://github.com/crossbario/autobahn-python) library is required for Web Sockets support. Refer to the [Autobahn installation](https://autobahn.readthedocs.io/en/latest/installation.html) for more details.
```

File: README.md
Comment (ccordoba12) [inline README.md]:
```suggestion
You can install this dependency with command below:
```

File: pylsp/python_lsp.py
Comment (ccordoba12) [inline pylsp/python_lsp.py]:
```suggestion
            """Handler to send responses of processed requests to the respective web socket clients."""
```

File: pylsp/python_lsp.py
Comment (ccordoba12) [inline pylsp/python_lsp.py]:
```suggestion
                # Binary message support will be provided in the future.
```

File: pylsp/python_lsp.py
Comment (ccordoba12) [inline pylsp/python_lsp.py]:
```suggestion
        """Entry point for consumer based server. Alternative to stream listeners."""
```

**Review by ccordoba12 — COMMENTED** [review]
@npradeep357, thanks for your patience with this! Last review and then this should be ready (we can take a look at adding tests later).

**Comment by npradeep357** [comment]
> @npradeep357, thanks for your patience with this! Last review and then this should be ready (we can take a look at adding tests later).

@ccordoba12 I'm planning to change the websockets library to reduce 3rd party library footprint. If it can wait a week or two, I will update the PR.

**Comment by ccordoba12** [comment]
Sure, no problem with that.

**Comment by npradeep357** [comment]
@ccordoba12, @krassowski, removed autobahn and twisted libraries and using websockets library to reduce libraries footprint.
please review, approve and merge this.

I have tested it my application and everything's working as expected..

websockets: https://websockets.readthedocs.io/en/stable/

**Comment by npradeep357** [comment]
@ccordoba12 @krassowski could you please review the latest changes.

**Comment by npradeep357** [comment]
@ccordoba12, can we merge this?

File: README.md
Comment (ccordoba12) [reply README.md]: Please address this suggestion.

File: pylsp/python_lsp.py
Comment (ccordoba12) [reply pylsp/python_lsp.py]: This suggestion is also still unaddressed. Please apply it.

**Review by ccordoba12 — COMMENTED** [review]
@npradeep357, thanks for your patience with this! There are still two unaddressed suggestions from my previous review. After you apply them, I'll merge your PR.

**Comment by npradeep357** [comment]
Hi @ccordoba12, Somehow these suggestions got overlooked. I have addressed them now. Thank you

**Review by ccordoba12 — APPROVED** [review]
Thanks for your contribution @npradeep357!

**Pull request merged**
